In [61]:
import pandas as pd
import numpy as np
import os
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error


In [62]:
#checking the path directory
print("Previous directory",os.getcwd())
#Now changing the path directory
path=r"C:\Users\Pavan\Desktop\Machine Learning Github\Home Works\HW1"
os.chdir(path)
print("Current directory",os.getcwd())

Previous directory C:\Users\Pavan\Desktop\Machine Learning Github\Home Works\HW1
Current directory C:\Users\Pavan\Desktop\Machine Learning Github\Home Works\HW1


Hosuing data set ----Question 1 
### Run linear regression using Normal Equations and Mean Square Error without using the libraries


In [63]:
#Loading the DataSet
col_names=["CRIM","ZN","INDUS","CHAS","NOX","RM","AGE","DIS","RAD","TAX","PTRATIO","B","LSTAT","MEDV"]
train_housing=pd.read_table("housing_train.txt",sep="\s+",names=col_names)
train_housing


    

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD    TAX  \
0     0.00632  18.0   2.31     0  0.538  6.575   65.2  4.0900    1  296.0   
1     0.02731   0.0   7.07     0  0.469  6.421   78.9  4.9671    2  242.0   
2     0.02729   0.0   7.07     0  0.469  7.185   61.1  4.9671    2  242.0   
3     0.03237   0.0   2.18     0  0.458  6.998   45.8  6.0622    3  222.0   
4     0.06905   0.0   2.18     0  0.458  7.147   54.2  6.0622    3  222.0   
5     0.02985   0.0   2.18     0  0.458  6.430   58.7  6.0622    3  222.0   
6     0.08829  12.5   7.87     0  0.524  6.012   66.6  5.5605    5  311.0   
7     0.14455  12.5   7.87     0  0.524  6.172   96.1  5.9505    5  311.0   
8     0.21124  12.5   7.87     0  0.524  5.631  100.0  6.0821    5  311.0   
9     0.17004  12.5   7.87     0  0.524  6.004   85.9  6.5921    5  311.0   
10    0.22489  12.5   7.87     0  0.524  6.377   94.3  6.3467    5  311.0   
11    0.11747  12.5   7.87     0  0.524  6.009   82.9  6.2267    5  311.0   
12    0.09378  12.5   7.87     0  0.524  5.889   39.0  5.4509    5  311.0   
13    0.62976   0.0   8.14     0  0.538  5.949   61.8  4.7075    4  307.0   
14    0.63796   0.0   8.14     0  0.538  6.096   84.5  4.4619    4  307.0   
15    0.62739   0.0   8.14     0  0.538  5.834   56.5  4.4986    4  307.0   
16    1.05393   0.0   8.14     0  0.538  5.935   29.3  4.4986    4  307.0   
17    0.78420   0.0   8.14     0  0.538  5.990   81.7  4.2579    4  307.0   
18    0.80271   0.0   8.14     0  0.538  5.456   36.6  3.7965    4  307.0   
19    0.72580   0.0   8.14     0  0.538  5.727   69.5  3.7965    4  307.0   
20    1.25179   0.0   8.14     0  0.538  5.570   98.1  3.7979    4  307.0   
21    0.85204   0.0   8.14     0  0.538  5.965   89.2  4.0123    4  307.0   
22    1.23247   0.0   8.14     0  0.538  6.142   91.7  3.9769    4  307.0   
23    0.98843   0.0   8.14     0  0.538  5.813  100.0  4.0952    4  307.0   
24    0.75026   0.0   8.14     0  0.538  5.924   94.1  4.3996    4  307.0   
25    1.13081   0.0   8.14     0  0.538  5.713   94.1  4.2330    4  307.0   
26    1.35472   0.0   8.14     0  0.538  6.072  100.0  4.1750    4  307.0   
27    1.38799   0.0   8.14     0  0.538  5.950   82.0  3.9900    4  307.0   
28    1.15172   0.0   8.14     0  0.538  5.701   95.0  3.7872    4  307.0   
29    1.61282   0.0   8.14     0  0.538  6.096   96.9  3.7598    4  307.0   
..        ...   ...    ...   ...    ...    ...    ...     ...  ...    ...   
403   4.03841   0.0  18.10     0  0.532  6.229   90.7  3.0993   24  666.0   
404   3.56868   0.0  18.10     0  0.580  6.437   75.0  2.8965   24  666.0   
405   4.64689   0.0  18.10     0  0.614  6.980   67.6  2.5329   24  666.0   
406   8.05579   0.0  18.10     0  0.584  5.427   95.4  2.4298   24  666.0   
407   6.39312   0.0  18.10     0  0.584  6.162   97.4  2.2060   24  666.0   
408   4.87141   0.0  18.10     0  0.614  6.484   93.6  2.3053   24  666.0   
409  15.02340   0.0  18.10     0  0.614  5.304   97.3  2.1007   24  666.0   
410  10.23300   0.0  18.10     0  0.614  6.185   96.7  2.1705   24  666.0   
411  14.33370   0.0  18.10     0  0.614  6.229   88.0  1.9512   24  666.0   
412   5.82401   0.0  18.10     0  0.532  6.242   64.7  3.4242   24  666.0   
413   5.70818   0.0  18.10     0  0.532  6.750   74.9  3.3317   24  666.0   
414   5.73116   0.0  18.10     0  0.532  7.061   77.0  3.4106   24  666.0   
415   2.81838   0.0  18.10     0  0.532  5.762   40.3  4.0983   24  666.0   
416   2.37857   0.0  18.10     0  0.583  5.871   41.9  3.7240   24  666.0   
417   3.67367   0.0  18.10     0  0.583  6.312   51.9  3.9917   24  666.0   
418   5.69175   0.0  18.10     0  0.583  6.114   79.8  3.5459   24  666.0   
419   4.83567   0.0  18.10     0  0.583  5.905   53.2  3.1523   24  666.0   
420   0.17331   0.0   9.69     0  0.585  5.707   54.0  2.3817    6  391.0   
421   0.27957   0.0   9.69     0  0.585  5.926   42.6  2.3817    6  391.0   
422   0.17899   0.0   9.69     0  0.585  5.670   28.8  2.7986    6  391.0   
423  

In [64]:
#This is the Y which is the output or the LABEL and Now I am separating the label from the dataset
y=train_housing["MEDV"]
y

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
5      28.7
6      22.9
7      27.1
8      16.5
9      18.9
10     15.0
11     18.9
12     21.7
13     20.4
14     18.2
15     19.9
16     23.1
17     17.5
18     20.2
19     18.2
20     13.6
21     19.6
22     15.2
23     14.5
24     15.6
25     12.7
26     14.5
27     13.2
28     13.1
29     13.5
       ... 
403    19.6
404    23.2
405    29.8
406    13.8
407    13.3
408    16.7
409    12.0
410    14.6
411    21.4
412    23.0
413    23.7
414    25.0
415    21.8
416    20.6
417    21.2
418    19.1
419    20.6
420    21.8
421    24.5
422    23.1
423    19.7
424    18.3
425    21.2
426    17.5
427    16.8
428    22.4
429    20.6
430    23.9
431    22.0
432    11.9
Name: MEDV, Length: 433, dtype: float64

In [65]:
#This is the FEATURE or the DATA MATRIX which conatins 13 features and y is the label
del train_housing["MEDV"]
train_housing

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD    TAX  \
0     0.00632  18.0   2.31     0  0.538  6.575   65.2  4.0900    1  296.0   
1     0.02731   0.0   7.07     0  0.469  6.421   78.9  4.9671    2  242.0   
2     0.02729   0.0   7.07     0  0.469  7.185   61.1  4.9671    2  242.0   
3     0.03237   0.0   2.18     0  0.458  6.998   45.8  6.0622    3  222.0   
4     0.06905   0.0   2.18     0  0.458  7.147   54.2  6.0622    3  222.0   
5     0.02985   0.0   2.18     0  0.458  6.430   58.7  6.0622    3  222.0   
6     0.08829  12.5   7.87     0  0.524  6.012   66.6  5.5605    5  311.0   
7     0.14455  12.5   7.87     0  0.524  6.172   96.1  5.9505    5  311.0   
8     0.21124  12.5   7.87     0  0.524  5.631  100.0  6.0821    5  311.0   
9     0.17004  12.5   7.87     0  0.524  6.004   85.9  6.5921    5  311.0   
10    0.22489  12.5   7.87     0  0.524  6.377   94.3  6.3467    5  311.0   
11    0.11747  12.5   7.87     0  0.524  6.009   82.9  6.2267    5  311.0   
12    0.09378  12.5   7.87     0  0.524  5.889   39.0  5.4509    5  311.0   
13    0.62976   0.0   8.14     0  0.538  5.949   61.8  4.7075    4  307.0   
14    0.63796   0.0   8.14     0  0.538  6.096   84.5  4.4619    4  307.0   
15    0.62739   0.0   8.14     0  0.538  5.834   56.5  4.4986    4  307.0   
16    1.05393   0.0   8.14     0  0.538  5.935   29.3  4.4986    4  307.0   
17    0.78420   0.0   8.14     0  0.538  5.990   81.7  4.2579    4  307.0   
18    0.80271   0.0   8.14     0  0.538  5.456   36.6  3.7965    4  307.0   
19    0.72580   0.0   8.14     0  0.538  5.727   69.5  3.7965    4  307.0   
20    1.25179   0.0   8.14     0  0.538  5.570   98.1  3.7979    4  307.0   
21    0.85204   0.0   8.14     0  0.538  5.965   89.2  4.0123    4  307.0   
22    1.23247   0.0   8.14     0  0.538  6.142   91.7  3.9769    4  307.0   
23    0.98843   0.0   8.14     0  0.538  5.813  100.0  4.0952    4  307.0   
24    0.75026   0.0   8.14     0  0.538  5.924   94.1  4.3996    4  307.0   
25    1.13081   0.0   8.14     0  0.538  5.713   94.1  4.2330    4  307.0   
26    1.35472   0.0   8.14     0  0.538  6.072  100.0  4.1750    4  307.0   
27    1.38799   0.0   8.14     0  0.538  5.950   82.0  3.9900    4  307.0   
28    1.15172   0.0   8.14     0  0.538  5.701   95.0  3.7872    4  307.0   
29    1.61282   0.0   8.14     0  0.538  6.096   96.9  3.7598    4  307.0   
..        ...   ...    ...   ...    ...    ...    ...     ...  ...    ...   
403   4.03841   0.0  18.10     0  0.532  6.229   90.7  3.0993   24  666.0   
404   3.56868   0.0  18.10     0  0.580  6.437   75.0  2.8965   24  666.0   
405   4.64689   0.0  18.10     0  0.614  6.980   67.6  2.5329   24  666.0   
406   8.05579   0.0  18.10     0  0.584  5.427   95.4  2.4298   24  666.0   
407   6.39312   0.0  18.10     0  0.584  6.162   97.4  2.2060   24  666.0   
408   4.87141   0.0  18.10     0  0.614  6.484   93.6  2.3053   24  666.0   
409  15.02340   0.0  18.10     0  0.614  5.304   97.3  2.1007   24  666.0   
410  10.23300   0.0  18.10     0  0.614  6.185   96.7  2.1705   24  666.0   
411  14.33370   0.0  18.10     0  0.614  6.229   88.0  1.9512   24  666.0   
412   5.82401   0.0  18.10     0  0.532  6.242   64.7  3.4242   24  666.0   
413   5.70818   0.0  18.10     0  0.532  6.750   74.9  3.3317   24  666.0   
414   5.73116   0.0  18.10     0  0.532  7.061   77.0  3.4106   24  666.0   
415   2.81838   0.0  18.10     0  0.532  5.762   40.3  4.0983   24  666.0   
416   2.37857   0.0  18.10     0  0.583  5.871   41.9  3.7240   24  666.0   
417   3.67367   0.0  18.10     0  0.583  6.312   51.9  3.9917   24  666.0   
418   5.69175   0.0  18.10     0  0.583  6.114   79.8  3.5459   24  666.0   
419   4.83567   0.0  18.10     0  0.583  5.905   53.2  3.1523   24  666.0   
420   0.17331   0.0   9.69     0  0.585  5.707   54.0  2.3817    6  391.0   
421   0.27957   0.0   9.69     0  0.585  5.926   42.6  2.3817    6  391.0   
422   0.17899   0.0   9.69     0  0.585  5.670   28.8  2.7986    6  391.0   
423  

In [66]:
#Using the Normal Equations I am calculating the coefficients of the paremets(w1,w2,w3-----w13)
#Now we have the Data matrix(X) and the labels(Y) now we need to find the coefficients
w_normal_equations=np.linalg.inv(train_housing.T.dot(train_housing)).dot(train_housing.T).dot(y)
w_normal_equations


array([-9.79342380e-02,  4.89586765e-02, -2.53928478e-02,  3.45087927e+00,
       -3.55458931e-01,  5.81653272e+00, -3.31447963e-03, -1.02050134e+00,
        2.26563208e-01, -1.22458785e-02, -3.88029879e-01,  1.70214971e-02,
       -4.85012955e-01])

In [67]:
w=np.reshape(w_normal_equations,(1,13))

In [68]:
##Function for calculating mean squared error using the Normal Equations for the training set
def mean_square_error(y,w,x,n_iterations):
    squared_error=0
    mean_squared_error=0
    for i in range(n_iterations):#For all the data points considerong the (Y-h(x))i.e(y-WX)
        squared_error=squared_error+((y[i]-(w.dot(x.iloc[i])))*(y[i]-(w.dot(x.iloc[i]))))
    mean_squared_error=squared_error/n_iterations
    
    print("Mean Squared Error for the training is:",mean_squared_error)

        
mean_square_error(y,w,train_housing,433)

Mean Squared Error for the training is: [24.47588278]


In [69]:
#Now we are conisdering the testing data set
test_data=pd.read_table("housing_test.txt",sep="\s+",header=None,names=col_names)
test_data

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE      DIS  RAD    TAX  \
0    0.84054   0.0   8.14     0  0.538  5.599   85.7   4.4546    4  307.0   
1    0.67191   0.0   8.14     0  0.538  5.813   90.3   4.6820    4  307.0   
2    0.95577   0.0   8.14     0  0.538  6.047   88.8   4.4534    4  307.0   
3    0.77299   0.0   8.14     0  0.538  6.495   94.4   4.4547    4  307.0   
4    1.00245   0.0   8.14     0  0.538  6.674   87.3   4.2390    4  307.0   
5    0.22927   0.0   6.91     0  0.448  6.030   85.5   5.6894    3  233.0   
6    0.25387   0.0   6.91     0  0.448  5.399   95.3   5.8700    3  233.0   
7    0.21977   0.0   6.91     0  0.448  5.602   62.0   6.0877    3  233.0   
8    0.10153   0.0  12.83     0  0.437  6.279   74.5   4.0522    5  398.0   
9    0.08707   0.0  12.83     0  0.437  6.140   45.8   4.0905    5  398.0   
10   0.05646   0.0  12.83     0  0.437  6.232   53.7   5.0141    5  398.0   
11   0.08387   0.0  12.83     0  0.437  5.874   36.6   4.5026    5  398.0   
12   0.04113  25.0   4.86     0  0.426  6.727   33.5   5.4007    4  281.0   
13   0.14866   0.0   8.56     0  0.520  6.727   79.9   2.7778    5  384.0   
14   0.11432   0.0   8.56     0  0.520  6.781   71.3   2.8561    5  384.0   
15   0.22876   0.0   8.56     0  0.520  6.405   85.4   2.7147    5  384.0   
16   0.21161   0.0   8.56     0  0.520  6.137   87.4   2.7147    5  384.0   
17   0.59005   0.0  21.89     0  0.624  6.372   97.9   2.3274    4  437.0   
18   0.32982   0.0  21.89     0  0.624  5.822   95.4   2.4699    4  437.0   
19   0.97617   0.0  21.89     0  0.624  5.757   98.4   2.3460    4  437.0   
20   0.55778   0.0  21.89     0  0.624  6.335   98.2   2.1107    4  437.0   
21   0.32264   0.0  21.89     0  0.624  5.942   93.5   1.9669    4  437.0   
22   2.15505   0.0  19.58     0  0.871  5.628  100.0   1.5166    5  403.0   
23   2.36862   0.0  19.58     0  0.871  4.926   95.7   1.4608    5  403.0   
24   1.80028   0.0  19.58     0  0.605  5.877   79.2   2.4259    5  403.0   
25   2.30040   0.0  19.58     0  0.605  6.319   96.1   2.1000    5  403.0   
26   2.44953   0.0  19.58     0  0.605  6.402   95.2   2.2625    5  403.0   
27   1.20742   0.0  19.58     0  0.605  5.875   94.6   2.4259    5  403.0   
28   0.12579  45.0   3.44     0  0.437  6.556   29.1   4.5667    5  398.0   
29   0.08370  45.0   3.44     0  0.437  7.185   38.9   4.5667    5  398.0   
..       ...   ...    ...   ...    ...    ...    ...      ...  ...    ...   
44   0.01965  80.0   1.76     0  0.385  6.230   31.5   9.0892    1  241.0   
45   0.03871  52.5   5.32     0  0.405  6.209   31.3   7.3172    6  293.0   
46   0.04301  80.0   1.91     0  0.413  5.663   21.9  10.5857    4  334.0   
47   0.10659  80.0   1.91     0  0.413  5.936   19.5  10.5857    4  334.0   
48   8.98296   0.0  18.10     1  0.770  6.212   97.4   2.1222   24  666.0   
49   3.84970   0.0  18.10     1  0.770  6.395   91.0   2.5052   24  666.0   
50   5.20177   0.0  18.10     1  0.770  6.127   83.4   2.7227   24  666.0   
51   4.26131   0.0  18.10     0  0.770  6.112   81.3   2.5091   24  666.0   
52  18.49820   0.0  18.10     0  0.668  4.138  100.0   1.1370   24  666.0   
53  19.60910   0.0  18.10     0  0.671  7.313   97.9   1.3163   24  666.0   
54  14.43830   0.0  18.10     0  0.597  6.852  100.0   1.4655   24  666.0   
55  51.13580   0.0  18.10     0  0.597  5.757  100.0   1.4130   24  666.0   
56  14.05070   0.0  18.10     0  0.597  6.657  100.0   1.5275   24  666.0   
57  18.81100   0.0  18.10     0  0.597  4.628  100.0   1.5539   24  666.0   
58  28.65580   0.0  18.10     0  0.597  5.155  100.0   1.5894   24  666.0   
59  12.24720   0.0  18.10     0  0.584  5.837   59.7   1.9976   24  666.0   
60  37.66190   0.0  18.10     0  0.679  6.202   78.7   1.8629   24  666.0   
61   7.36711   0.0  18.10     0  0.679  6.193   78.1   1.9356   24  666.0   
62  10.67180   0.0  18.10     0  0.740  6.459   94.8   1.9879   24  666.0   
63   6.28807   0.0  18.10     0  0.740  6.341   96.4   2.0720   24  666.0   
64  

In [70]:
y_test=test_data["MEDV"]
y_test

0     13.9
1     16.6
2     14.8
3     18.4
4     21.0
5     16.6
6     14.4
7     19.4
8     20.0
9     20.8
10    21.2
11    20.3
12    28.0
13    27.5
14    26.5
15    18.6
16    19.3
17    23.0
18    18.4
19    15.6
20    18.1
21    17.4
22    15.6
23    14.6
24    23.8
25    23.8
26    22.3
27    17.4
28    29.8
29    34.9
      ... 
44    20.1
45    23.2
46    18.2
47    20.6
48    17.8
49    21.7
50    22.7
51    22.6
52    13.8
53    15.0
54    27.5
55    15.0
56    17.2
57    17.9
58    16.3
59    10.2
60    10.9
61    11.0
62    11.8
63    14.9
64    12.6
65    19.9
66    19.0
67    19.1
68    19.1
69    13.6
70    20.1
71    21.8
72    24.5
73    23.1
Name: MEDV, Length: 74, dtype: float64

In [71]:
#Here we are deleting the label Y from the Data/feature matrix
del test_data["MEDV"]


In [72]:
#Now finding the error with respect to the test data
x_test=test_data
x_test

CRIM    ZN  INDUS  CHAS    NOX     RM    AGE      DIS  RAD    TAX  \
0    0.84054   0.0   8.14     0  0.538  5.599   85.7   4.4546    4  307.0   
1    0.67191   0.0   8.14     0  0.538  5.813   90.3   4.6820    4  307.0   
2    0.95577   0.0   8.14     0  0.538  6.047   88.8   4.4534    4  307.0   
3    0.77299   0.0   8.14     0  0.538  6.495   94.4   4.4547    4  307.0   
4    1.00245   0.0   8.14     0  0.538  6.674   87.3   4.2390    4  307.0   
5    0.22927   0.0   6.91     0  0.448  6.030   85.5   5.6894    3  233.0   
6    0.25387   0.0   6.91     0  0.448  5.399   95.3   5.8700    3  233.0   
7    0.21977   0.0   6.91     0  0.448  5.602   62.0   6.0877    3  233.0   
8    0.10153   0.0  12.83     0  0.437  6.279   74.5   4.0522    5  398.0   
9    0.08707   0.0  12.83     0  0.437  6.140   45.8   4.0905    5  398.0   
10   0.05646   0.0  12.83     0  0.437  6.232   53.7   5.0141    5  398.0   
11   0.08387   0.0  12.83     0  0.437  5.874   36.6   4.5026    5  398.0   
12   0.04113  25.0   4.86     0  0.426  6.727   33.5   5.4007    4  281.0   
13   0.14866   0.0   8.56     0  0.520  6.727   79.9   2.7778    5  384.0   
14   0.11432   0.0   8.56     0  0.520  6.781   71.3   2.8561    5  384.0   
15   0.22876   0.0   8.56     0  0.520  6.405   85.4   2.7147    5  384.0   
16   0.21161   0.0   8.56     0  0.520  6.137   87.4   2.7147    5  384.0   
17   0.59005   0.0  21.89     0  0.624  6.372   97.9   2.3274    4  437.0   
18   0.32982   0.0  21.89     0  0.624  5.822   95.4   2.4699    4  437.0   
19   0.97617   0.0  21.89     0  0.624  5.757   98.4   2.3460    4  437.0   
20   0.55778   0.0  21.89     0  0.624  6.335   98.2   2.1107    4  437.0   
21   0.32264   0.0  21.89     0  0.624  5.942   93.5   1.9669    4  437.0   
22   2.15505   0.0  19.58     0  0.871  5.628  100.0   1.5166    5  403.0   
23   2.36862   0.0  19.58     0  0.871  4.926   95.7   1.4608    5  403.0   
24   1.80028   0.0  19.58     0  0.605  5.877   79.2   2.4259    5  403.0   
25   2.30040   0.0  19.58     0  0.605  6.319   96.1   2.1000    5  403.0   
26   2.44953   0.0  19.58     0  0.605  6.402   95.2   2.2625    5  403.0   
27   1.20742   0.0  19.58     0  0.605  5.875   94.6   2.4259    5  403.0   
28   0.12579  45.0   3.44     0  0.437  6.556   29.1   4.5667    5  398.0   
29   0.08370  45.0   3.44     0  0.437  7.185   38.9   4.5667    5  398.0   
..       ...   ...    ...   ...    ...    ...    ...      ...  ...    ...   
44   0.01965  80.0   1.76     0  0.385  6.230   31.5   9.0892    1  241.0   
45   0.03871  52.5   5.32     0  0.405  6.209   31.3   7.3172    6  293.0   
46   0.04301  80.0   1.91     0  0.413  5.663   21.9  10.5857    4  334.0   
47   0.10659  80.0   1.91     0  0.413  5.936   19.5  10.5857    4  334.0   
48   8.98296   0.0  18.10     1  0.770  6.212   97.4   2.1222   24  666.0   
49   3.84970   0.0  18.10     1  0.770  6.395   91.0   2.5052   24  666.0   
50   5.20177   0.0  18.10     1  0.770  6.127   83.4   2.7227   24  666.0   
51   4.26131   0.0  18.10     0  0.770  6.112   81.3   2.5091   24  666.0   
52  18.49820   0.0  18.10     0  0.668  4.138  100.0   1.1370   24  666.0   
53  19.60910   0.0  18.10     0  0.671  7.313   97.9   1.3163   24  666.0   
54  14.43830   0.0  18.10     0  0.597  6.852  100.0   1.4655   24  666.0   
55  51.13580   0.0  18.10     0  0.597  5.757  100.0   1.4130   24  666.0   
56  14.05070   0.0  18.10     0  0.597  6.657  100.0   1.5275   24  666.0   
57  18.81100   0.0  18.10     0  0.597  4.628  100.0   1.5539   24  666.0   
58  28.65580   0.0  18.10     0  0.597  5.155  100.0   1.5894   24  666.0   
59  12.24720   0.0  18.10     0  0.584  5.837   59.7   1.9976   24  666.0   
60  37.66190   0.0  18.10     0  0.679  6.202   78.7   1.8629   24  666.0   
61   7.36711   0.0  18.10     0  0.679  6.193   78.1   1.9356   24  666.0   
62  10.67180   0.0  18.10     0  0.740  6.459   94.8   1.9879   24  666.0   
63   6.28807   0.0  18.10     0  0.740  6.341   96.4   2.0720   24  666.0   
64  

In [73]:
##The mean squared error function is already defined before and I just called that function using the function name
mean_square_error(y_test,w,x_test,74)

Mean Squared Error for the training is: [24.29223818]


<b>train error: 24.47<br>
<b>test error: 24.29<br>
<b>using the Normal Equations

<b>Part 2 : Perform everything using the ScikitLearn package<br>

In [74]:
##Mean squared error using Scikit Learn for the training data
lin_reg = LinearRegression()
lin_reg.fit(train_housing, y)# Finding the parameters
lin_reg.intercept_, lin_reg.coef_
train_predictions=lin_reg.predict(train_housing)
print("\n The MSE Error for testing is:",mean_squared_error(y,train_predictions))



 The MSE Error for testing is: 22.081273187013167


In [75]:
##Now calculating the mean sqaured error for the testing dataset
test_predictions=lin_reg.predict(test_data)
print("\n The MSE Error for the testing is:",mean_squared_error(y_test,test_predictions))


 The MSE Error for the testing is: 22.638256296589123


<b> Training error: 22.08<br>
<b> Testing error:  22.63<br>
<b> using the scikitlear package